### PR TITLE
add BiLSTM and attention layer

### PR DESCRIPTION
**Pull request by afaji**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -103,12 +110,26 @@ def __init__(self,
             ])
             total_features_size += sum(suffix_embedding_size)
 
+        ff_input_size = total_features_size
+        
+        self.lstm = None
+        if (True):
```
Comment (kmkurn) [inline models.py]: Shouldn't this be `if use_lstm:`?

File: models.py [hunk 2]
```diff
@@ -103,12 +110,26 @@ def __init__(self,
...
+            self.lstm = nn.LSTM(total_features_size, hidden_size,
+                                num_layers=2, bidirectional=True, batch_first=True)
+            ff_input_size = 2 * hidden_size
+        
+        self.attention = None
+        if (True):
```
Comment (kmkurn) [inline models.py]: Shouldn't this be `if use_attention:`?

File: models.py [hunk 3]
```diff
@@ -103,12 +110,26 @@ def __init__(self,
...
+            self.attention = Attention(ff_input_size, attention_type = 'general')
```
Comment (kmkurn) [inline models.py]: Please discard spaces surrounding the equal sign in keyword arguments (i.e. `attention_type='general'`)

File: models.py [hunk 4]
```diff
@@ -192,6 +221,9 @@ def reset_parameters(self) -> None:
             for emb in self.suffix_embedding:
                 emb.reset_parameters()
 
+        if self.uses_lstm:
+            self.lstm.reset_parameters()
+
```
Comment (kmkurn) [inline models.py:214]: How about the attention's parameters? Shouldn't they be initialized here as well?

File: models.py [hunk 5]
```diff
@@ -103,12 +110,26 @@ def __init__(self,
...
+        self.attention = None
+        if (True):
+            self.attention = Attention(ff_input_size, attention_type = 'general')
+            ff_input_size *= 2
+         
         self.ff = nn.Sequential(
-            nn.Linear(total_features_size, hidden_size),
+            nn.Linear(ff_input_size, hidden_size),
             nn.Tanh(),
             nn.Dropout(dropout),
```
Comment (kmkurn) [inline models.py:121]: Should the dropout layer be removed here because we're following Lample et al.'s? Do they use dropout here as well?

File: models.py [hunk 6]
```diff
@@ -316,12 +357,29 @@ def _compute_emissions(self,
...
 
-        # shape: (batch_size, seq_length, (2*window + 1) * word_embedding_size)
-        inputs = word_features
         if affix_features is not None:
             # shape: (batch_size, seq_length, (2*window+1)*word_embedding_size+total_affix_emb_size)
             inputs = torch.cat((inputs, affix_features), dim=-1)
 
+        # apply Bidirectional LSTM
+        if self.uses_lstm:
+            seq_lengths = torch.sum(mask.int(),dim = 1)
```
Comment (kmkurn) [inline models.py]: 1. I think the case when `mask` is `None` has not been handled, has it?
1. Please put a space after comma (and discard spaces around the equal sign).

File: models.py [hunk 7]
```diff
@@ -316,12 +357,29 @@ def _compute_emissions(self,
...
+            seq_lengths, sent_perm = seq_lengths.sort(0, descending=True)
+            inputs = inputs[sent_perm]
+
+            packed_input = pack_padded_sequence(inputs, seq_lengths.cpu().data.numpy(), batch_first=True)
```
Comment (kmkurn) [inline models.py]: Can't this work without moving `seq_lengths` to CPU nor converting it to numpy array? e.g. just `seq_lengths.data`.

File: models.py [hunk 8]
```diff
@@ -3,13 +3,16 @@
...
+from torch.nn.utils.rnn import pad_packed_sequence, pack_padded_sequence
 from torch.autograd import Variable as Var
 from torchcrf import CRF
 import torch
 import torch.nn as nn
+
 import torch.nn.functional as F
 import torch.nn.init as init
 
+from torchnlp.nn import Attention
```
Comment (kmkurn) [inline models.py]: Add `pytorch-nlp` as a dependency in `environment.yml` please.

File: models.py [hunk 9]
```diff
@@ -285,25 +318,33 @@ def _compute_emissions(self,
         assert prefixes is None or prefixes.size() == (batch_size, seq_length, 2)
         assert suffixes is None or suffixes.size() == (batch_size, seq_length, 2)
```
Comment (kmkurn) [inline models.py:308]: Put asserts for `mask` after this line please.

**Review by kmkurn — CHANGES_REQUESTED** [review]
See inline comments for details. There are minor stylistic issues; you should run `flake8` linter from the project root to check for such issues (and more).

File: models.py (hunk 9 above)
Comment (afaji) [reply models.py:308]: added

File: models.py (hunk 5 above)
Comment (afaji) [reply models.py:121]: Lample et al.'s did not use a dropout layer here, this seems to help reduce overfitting.

File: models.py (hunk 4 above)
Comment (afaji) [reply models.py:214]: attention is removed for time being

File: models.py [hunk 10]
```diff
@@ -363,23 +352,22 @@ def _compute_emissions(self,
 
         # apply Bidirectional LSTM
         if self.uses_lstm:
-            seq_lengths = torch.sum(mask.int(),dim = 1)
+            if mask is None:
+                # shape (batch_size, seq_length)
+                mask = self._get_mask_for(words)   
+            
+            seq_lengths = torch.sum(mask.int(), dim=1)
```
Comment (kmkurn) [inline models.py]: Could you please add `# shape:` comments before any lines whose LHS is a new tensor? (e.g. this line, line 366, etc.) It's rather hard to keep track of all the shapes without such comments.

File: models.py [hunk 11]
```diff
@@ -363,23 +352,22 @@ def _compute_emissions(self,
...
             seq_lengths, sent_perm = seq_lengths.sort(0, descending=True)
             inputs = inputs[sent_perm]
 
-            packed_input = pack_padded_sequence(inputs, seq_lengths.cpu().data.numpy(), batch_first=True)
-     
+            packed_input = pack_padded_sequence(inputs, seq_lengths.data.tolist(), batch_first=True)
```
Comment (kmkurn) [inline models.py]: Can just `seq_lengths.data` work? Calling `.tolist()` is *very* slow, up to 1000x times slower than `.numpy()`. If just `.data` works then use that, no need to call `.numpy()`. If it doesn't, then `.numpy()` is preferred as it's faster.

File: models.py (hunk 5 above)
Comment (kmkurn) [reply models.py]: I see, so we prefer to keep this here.

File: models.py [hunk 12]
```diff
@@ -316,12 +346,28 @@ def _compute_emissions(self,
...
+        if self.uses_lstm:
+            if mask is None:
+                # shape (batch_size, seq_length)
+                mask = self._get_mask_for(words)   
+            
+            seq_lengths = torch.sum(mask.int(), dim=1)
+            seq_lengths, sent_perm = seq_lengths.sort(0, descending=True)
+            inputs = inputs[sent_perm]
+
+            packed_input = pack_padded_sequence(inputs, seq_lengths.data.tolist(), batch_first=True)
```
Comment (kmkurn) [inline models.py]: Can just `.data` works? Calling `.tolist()` is *very* slow, up to 1000x times slower than `.numpy()`. If just `.data` works, use that and no need to call `.numpy()`. If it doesn't, then `.numpy()` is preferred as it's faster.

File: models.py [hunk 13]
```diff
@@ -277,33 +297,43 @@ def _check_dims_and_sizes(self,
...
-        for i in range(seq_length):
-            lo, hi = i, i + 2 * self.window + 1
-            # shape: (batch_size, (2*window + 1) * word_embedding_size)
-            word_feats = embedded_words[:, lo:hi, :].contiguous().view(batch_size, -1)
-            temp.append(word_feats)
-        # shape: (batch_size, seq_length, (2*window + 1) * word_embedding_size)
-        word_features = torch.stack(temp, dim=1)
+        embedded_words = self.dropout(embedded_words)
+        embedded_prefixes = self.dropout(embedded_prefixes)
+        embedded_suffixes = self.dropout(embedded_suffixes)
```
Comment (kmkurn) [inline models.py]: Line 324 and 325 will raise errors if `use_prefix=False` or `use_suffix=False` respectively. This is because `embedded_prefixes/suffixes` will be `None` when the corresponding flag is `False`. Could you handle this?

**Review by kmkurn — CHANGES_REQUESTED** [review]

**Review by kmkurn — APPROVED** [review]
LGTM. I'll merge this.
